contracts/TokenListManager.sol

```diff
@@ -6,176 +6,80 @@ import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+    // Mappings for whitelisted and blacklisted tokens, categorized by token type.
+    mapping(TokenType => EnumerableSet.AddressSet) private whitelistedTokens;
+    mapping(TokenType => EnumerableSet.AddressSet) private blacklistedTokens;
+
+    // Contract constructor setting up initial admin role.
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(USER_ROLE, ADMIN_ROLE);
     }
 
-    // Implement the addTokens method and add tokens to whitelist
-    function addTokens(address[] calldata tokens) external override onlyOwner {
-        for (uint i = 0; i < tokens.length; i++) {
-            // Check if the token is not already whitelisted
-            if (!whitelistedTokens[tokens[i]]) {
-                // If the token type is ERC-20
-                if (IERC165(tokens[i]).supportsInterface(type(IERC20).interfaceId)) {
-                    // Storage to whitelistErc20Tokens
-                    whitelistErc20Tokens.push(tokens[i]);
-                // If the token type is IERC-721
-                } else if (IERC165(tokens[i]).supportsInterface(type(IERC721).interfaceId)) {
-                    // Storage to whitelistErc721Tokens
-                    whitelistErc721Tokens.push(tokens[i]);
-                // If the token type is IERC-1155
-                } else if (IERC165(tokens[i]).supportsInterface(type(IERC1155).interfaceId)) {
-                    // Storage to whitelistErc1155Tokens
-                    whitelistErc1155Tokens.push(tokens[i]);
-                } else {
-                    revert("Unsupported token type");
-                }
-                // Mark the token address as whitelisted in the mapping
-                whitelistedTokens[tokens[i]] = true;
-                // Add the token address to the array of whitelisted tokens
-                whitelistedTokensArray.push(tokens[i]);
-            }
-        }
+    // Modifier to restrict function access to users with the admin role.
+    modifier onlyAdmin() {
+        require(hasRole(ADMIN_ROLE, msg.sender), "Not authorized: caller is not an admin");
+        _;
     }
 
-    // Implement the removeTokens method and remove tokens from whitelist
-    function removeTokens(address[] calldata tokens) external override onlyOwner {
+    // Implement the addTokens method and add tokens to whitelist, only accessible by admin
+    function addTokens(address[] calldata tokens, TokenType tokenType) external onlyAdmin {
         for (uint i = 0; i < tokens.length; i++) {
-            // Check if the token in whitelist
-            if (whitelistedTokens[tokens[i]]) {
-                // Remove each token address from whitelist
-                whitelistedTokens[tokens[i]] = false;
-                // Call internal function below to remove from the array
-                removeTokenFromWhitelistArray(tokens[i]);
-            }
+            whitelistedTokens[tokenType].add(tokens[i]);
```

Comment:
What if the token is already blacklisted?

Reply:
Solved